### PR TITLE
PP API and nD Distributed Timeline Profiling

### PR DESCRIPTION
**Pull request by MackZackA**

### CODE REVIEW

File: examples/llama2_4D_finetune/data/shakespeare/prepare.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 # SOFTWARE.
 ################################################################################
-# Modification Copyright 2023 ByteDance Ltd. and/or its affiliates.
+# Modification Copyright 2024 ByteDance Ltd. and/or its affiliates.
```
Comment (leonardo0lyj) [inline examples/llama2_4D_finetune/data/shakespeare/prepare.py:22]: - 2023 is correct; the first date of vescale;
- 2024 is wrong

File: docs/pictures/pp.png
Comment (leonardo0lyj) [inline docs/pictures/pp.png:1]: fix the figure as suggested in internal repo

File: vescale/pipe/README.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    model = EightMLP()
+    # or model = deferred_init(EightMLP) 
+
+    from vescale.plan import PipelineParallelPlan, PipelineScheduleType, PipelineSplitMethodType, ModeType
+    from vescale.pipe import construct_pipeline_stage
+    from vescale.engine import PipeEngine
+    from vescale.dtensor.device_mesh import DeviceMesh
+    
+    # create 3-dim DeviceMesh
+    device_mesh = DeviceMesh("cuda", [[[0]], [[1]], [[2]], [[3]]], mesh_dim_names=("PP", "DP", "TP"))
```
Comment (leonardo0lyj) [inline vescale/pipe/README.md:93]: why not fixed? why we need 3D mesh to run pp?

**Comment by MingjiHan99** [comment]
LGTM

**Review by jc-bytedance — APPROVED** [review]

File: examples/open_llama_4D_benchmark/download_open_llama_ckpt.py [hunk 3]
```diff
@@ -1,6 +1,6 @@
 ################################################################################
 #
-# Copyright 2023 ByteDance Ltd. and/or its affiliates. All rights reserved.
+# Copyright 2024 ByteDance Ltd. and/or its affiliates. All rights reserved.
```
Comment (leonardo0lyj) [inline examples/open_llama_4D_benchmark/download_open_llama_ckpt.py:3]: @MackZackA ？

File: examples/open_llama_4D_benchmark/llama_mfu_calculator.py [hunk 4]
```diff
@@ -1,6 +1,6 @@
 ################################################################################
 #
-# Copyright 2023 ByteDance Ltd. and/or its affiliates. All rights reserved.
+# Copyright 2024 ByteDance Ltd. and/or its affiliates. All rights reserved.
```
Comment (leonardo0lyj) [inline examples/open_llama_4D_benchmark/llama_mfu_calculator.py:3]: ？

File: examples/open_llama_4D_benchmark/run_open_llama_w_vescale.py [hunk 5]
```diff
@@ -1,6 +1,6 @@
 ################################################################################
 #
-# Copyright 2023 ByteDance Ltd. and/or its affiliates. All rights reserved.
+# Copyright 2024 ByteDance Ltd. and/or its affiliates. All rights reserved.
```
Comment (leonardo0lyj) [inline examples/open_llama_4D_benchmark/run_open_llama_w_vescale.py:3]: ？

File: examples/open_llama_4D_benchmark/sharding_plan.py [hunk 6]
```diff
@@ -1,6 +1,6 @@
 ################################################################################
 #
-# Copyright 2023 ByteDance Ltd. and/or its affiliates. All rights reserved.
+# Copyright 2024 ByteDance Ltd. and/or its affiliates. All rights reserved.
```
Comment (leonardo0lyj) [inline examples/open_llama_4D_benchmark/sharding_plan.py:3]: ？

File: test/checkpoint/open_llama/test_open_llama_dp_reshard.py [hunk 7]
```diff
@@ -1,6 +1,6 @@
 ################################################################################
 #
-# Copyright 2023 ByteDance Ltd. and/or its affiliates. All rights reserved.
+# Copyright 2024 ByteDance Ltd. and/or its affiliates. All rights reserved.
```
Comment (leonardo0lyj) [inline test/checkpoint/open_llama/test_open_llama_dp_reshard.py:3]: ？

File: test/checkpoint/open_llama/test_open_llama_load_save.py [hunk 8]
```diff
@@ -1,6 +1,6 @@
 ################################################################################
 #
-# Copyright 2023 ByteDance Ltd. and/or its affiliates. All rights reserved.
+# Copyright 2024 ByteDance Ltd. and/or its affiliates. All rights reserved.
```
Comment (leonardo0lyj) [inline test/checkpoint/open_llama/test_open_llama_load_save.py:3]: ？

File: test/checkpoint/open_llama/test_open_llama_tp_reshard.py [hunk 9]
```diff
@@ -1,6 +1,6 @@
 ################################################################################
 #
-# Copyright 2023 ByteDance Ltd. and/or its affiliates. All rights reserved.
+# Copyright 2024 ByteDance Ltd. and/or its affiliates. All rights reserved.
```
Comment (leonardo0lyj) [inline test/checkpoint/open_llama/test_open_llama_tp_reshard.py:3]: ？

File: test/parallel/pipeline/e2e/test_pp_accuracy_alignment.py [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+        optimizer = torch.optim.SGD(_parameters, **optimizer_fn_kwargs)
+        pipe_module = PipeModule(stage_modules, optimizer, None, stage_dependency, p2p_index_mapping, pipe_config)
+        engine = PipeEngine(
+            pipe_module,
+            VESCALE_DEVICE_MESH,
+            self.loss_fn,
+            pipe_config,
+        )
+
+        engine.forward_backward(batch)
```
Comment (leonardo0lyj) [inline test/parallel/pipeline/e2e/test_pp_accuracy_alignment.py:228]: ？

File: vescale/engine/pipe.py [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+import os
+
+
+class PipeEngine:
+    def __init__(
+        self,
+        module: PipeModule,
+        global_mesh: VeDeviceMesh,
+        loss_fn: Callable,
+        config: PipelineParallelPlan,
```
Comment (leonardo0lyj) [inline vescale/engine/pipe.py:39]: `plan`?

File: vescale/model/base_gpt/__init__.py [hunk 12]
```diff
@@ -0,0 +1,5 @@
+################################################################################
```
Comment (leonardo0lyj) [inline vescale/model/base_gpt/__init__.py:1]: this `base_gpt` should not be open sourced

**Comment by leonardo0lyj** [comment]
+ main README.md update

File: examples/open_llama_4D_benchmark/download_open_llama_ckpt.py (hunk 3 above)
Comment (WillemJiang) [reply examples/open_llama_4D_benchmark/download_open_llama_ckpt.py:3]: We normally update the year of copyright when made some new changes that year.

**Pull request merged**